src/ring_theory/graded_algebra/radical.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import ring_theory.graded_algebra.homogeneous_ideal
+
+/-!
+
+This file contains a proof that the radical of any homogeneous ideal is a homogeneous ideal
+
+## Main statements
+
+* `homogeneous_ideal.is_prime_iff`: a homogeneous ideal `I` is prime if and only if `I` is
+  homogeneously prime, i.e. if `x, y` are homogeneous elements such that `x * y ∈ I`, then
+  at least one of `x,y` is in `I`.
+* `homogeneous_ideal.rad`: radical of homogeneous ideal is a homogeneous ideal.
+
+## Implementation details
+
+Throughout this file, the indexing type `ι` of grading is assumed to be a
+`linear_ordered_cancel_add_comm_monoid`. This might be stronger than necessary and `linarith`
+does not work on `linear_ordered_cancel_add_comm_monoid`.
+
+## Tags
+
+homogeneous, radical
+-/
+
+open graded_algebra set_like finset
+open_locale big_operators
+
+variables {ι R A : Type*}
+variables [comm_semiring R] [comm_ring A] [algebra R A]
+variables [linear_ordered_cancel_add_comm_monoid ι] [decidable_eq ι]
+variables {𝒜 : ι → submodule R A} [graded_algebra 𝒜]
+
+lemma ideal.is_homogeneous.is_prime_of_homogeneous_mem_or_mem
+  {I : ideal A} (hI : I.is_homogeneous 𝒜) (I_ne_top : I ≠ ⊤)
+  (homogeneous_mem_or_mem : ∀ {x y : A},
+    is_homogeneous 𝒜 x → is_homogeneous 𝒜 y → (x * y ∈ I → x ∈ I ∨ y ∈ I)) :
+  ideal.is_prime I :=
+⟨I_ne_top, begin
+  intros x y hxy, by_contradiction rid,
+  obtain ⟨rid₁, rid₂⟩ := not_or_distrib.mp rid,
+  -- the idea of the proof is the following :
+  -- since `x * y ∈ I` and `I` homogeneous, then `proj i (x * y) ∈ I` for any `i : ι`.
+  -- Then consider two sets `{i ∈ x.support | xᵢ ∉ I}` and `{j ∈ y.support | yⱼ ∉ J}`;
+  -- let `max₁, max₂` be the maximum of the two sets, then `proj (max₁ + max₂) (x * y) ∈ I`.
+  -- Then, `proj max₁ x ∉ I` and `proj max₂ j ∉ I`
+  -- but `proj i x ∈ I` for all `max₁ < i` and `proj j y ∈ I` for all `max₂ < j`.
+  -- `  proj (max₁ + max₂) (x * y)`
+  -- `= ∑ {(i, j) ∈ supports | i + j = max₁ + max₂}, xᵢ * yⱼ`
+  -- `= proj max₁ x * proj max₂ y`
+  -- `  + ∑ {(i, j) ∈ supports \ {(max₁, max₂)} | i + j = max₁ + max₂}, xᵢ * yⱼ`.
+  -- This is a contradiction, because both `proj (max₁ + max₂) (x * y) ∈ I` and the sum on the
+  -- right hand side is in `I` however `proj max₁ x * proj max₂ y` is not in `I`.
```

Comment:
```suggestion
  /-
  The idea of the proof is the following :
  since `x * y ∈ I` and `I` homogeneous, then `proj i (x * y) ∈ I` for any `i : ι`.
  Then consider two sets `{i ∈ x.support | xᵢ ∉ I}` and `{j ∈ y.support | yⱼ ∉ I}`;
  let `max₁, max₂` be the maximum of the two sets, then `proj (max₁ + max₂) (x * y) ∈ I`.
  Then, `proj max₁ x ∉ I` and `proj max₂ j ∉ I`
  but `proj i x ∈ I` for all `max₁ < i` and `proj j y ∈ I` for all `max₂ < j`.

    proj (max₁ + max₂) (x * y)
  = ∑ {(i, j) ∈ supports | i + j = max₁ + max₂}, xᵢ * yⱼ`
  = proj max₁ x * proj max₂ y`
    + ∑ {(i, j) ∈ supports \ {(max₁, max₂)} | i + j = max₁ + max₂}, xᵢ * yⱼ`.

  This is a contradiction, because both `proj (max₁ + max₂) (x * y) ∈ I` and the sum on the
  right hand side is in `I` however `proj max₁ x * proj max₂ y` is not in `I`.
  -/
```

Reply:
>  however `proj max₁ x * proj max₂ y` is not in `I`

Could you extract this statement to a `have` somewhere?

Reply:
Done!